### PR TITLE
Input widget drop down

### PR DESCRIPTION
**Pull request by robinr**
InputDropDown Widget provides a capability to choose the input based on the following two methods
a) Entry of the complete name present in the option 
b) Selecting one of the options present in the datalist.

Note: 
All the other remaining features of InputWidget is inherited except for the "spellcheck" (true or false), a new prop is provided to either enable or disable the invalid feed back for dropdown "message_if_blank"

### CODE REVIEW

File: components/playback/controls/player/InputDropDown.vue [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+      input_width_background: this.input_width + 4,
+    };
+  },
+  computed: {
+    input_key_action: function () {
+      return this.message_if_blank;
+    },
+    dropdown_options: function () {
+      if (this.options_text.length == 0) {
+        return [];
```
Comment (eli88fine) [inline components/playback/controls/player/InputDropDown.vue:100]: please make sure this line/logic is tested. In test driven development, we don't write a line of code without first having a failing test.   You could test it as part of a test `Given the component is mounted with no options, When the prop options_text is updated by the parent and the user types one of the new valid options, Then an update event is emitted with the typed option`.   This then also tests to confirm the component reacts appropriately to a prop update---since there is a lot of logic between updating that prop and the desired final result of getting the items into the `datalist` with the right ids, it's not a bad idea to confirm it is all wired up correctly.

**Review by eli88fine — CHANGES_REQUESTED** [review]
Your variable naming, unit tests, and E2E tests are all very much improved from even just a few weeks ago. Great work!  There are some minor corrections, one point of functionality missing, and a visual correction to the alignment needed.

File: pages/playback/controls/player/input-dropdown/input-dropdown-disallow.vue [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      disallow_entry: true,
+      nicknames_list: [
+        "Customer Account 1",
+        "Customer Account-2",
+        "Customer Account-3",
+      ],
+      on_empty_flag: false,
+    };
+  },
+  watch: {
```
[the remark this answers is not in the document]
Comment (robinr) [reply pages/playback/controls/player/input-dropdown/input-dropdown-disallow.vue:39]: Agreed

File: pages/playback/controls/player/input-dropdown/basic-dropdown.vue [hunk 3]
```diff
@@ -0,0 +1,56 @@
+<template>
```
[the remark this answers is not in the document]
Comment (robinr) [reply pages/playback/controls/player/input-dropdown/basic-dropdown.vue:1]: Agreed, will provide an example.

File: tests/e2e/playback/controls/player/input-dropdown-e2e.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    "input-dropdown"
+  );
+  const screenshot_path = path.join(
+    screenshot_path_base,
+    "input-dropdown-no-value"
+  );
+  await testcafe_page_visual_regression(t, screenshot_path);
+});
+
+test("testing the Input DropDown for the VALID Value Entered", async (t) => {
```
[the remark this answers is not in the document]
Comment (robinr) [reply tests/e2e/playback/controls/player/input-dropdown-e2e.js:28]: ok will implement a part of the parent function, without having the button controls, will work on the them.

File: tests/e2e/playback/controls/player/input-dropdown-e2e.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  );
+  const screenshot_path = path.join(screenshot_path_base, "x-y-offset");
+  await testcafe_page_visual_regression(t, screenshot_path);
+});
+
+fixture`playback/controls/player/input-dropdown/input-dropdown-width-200`
+  .page // declare the fixture
+`http://localhost:8080/playback/controls/player/input-dropdown/input-dropdown-width-200`; // specify the start page
+
+test("testing the Input DropDown for width of 200px", async (t) => {
```
[the remark this answers is not in the document]
Comment (robinr) [reply tests/e2e/playback/controls/player/input-dropdown-e2e.js:63]: Would work on the centre issue.. need some correction on calculation based on the number of alphabets in the title label. will fix the issue.

File: tests/e2e/playback/controls/player/input-dropdown-e2e.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+  await t.typeText(input_field, "sample text");
+  await t.click(input_label);
+  await testcafe_page_visual_regression(t, screenshot_path);
+});
+
+fixture`playback/controls/player/input-dropdown/input-no-title-dropdown`
+  .page // declare the fixture
+`http://localhost:8080/playback/controls/player/input-dropdown/input-no-title-dropdown`; // specify the start page
+
+test("testing the Input Widget when the input is disabled or not allowed then entering text doesn't update visually", async (t) => {
```
[the remark this answers is not in the document]
Comment (robinr) [reply tests/e2e/playback/controls/player/input-dropdown-e2e.js:119]: Agreed, needs correction.

File: tests/unit/components/playback/controls/InputDropDown.spec.js [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+    wrapper = mount(ComponentToTest, {
+      propsData,
+      store,
+      localVue,
+    });
+    const userdata = "Demo Account"; // some data
+    const input_widget = wrapper.find("#input-dropdown-widget");
+    input_widget.element.value = userdata;
+    await input_widget.trigger("input");
+    expect(input_widget.html()).toContain("disabled");
```
[the remark this answers is not in the document]
Comment (robinr) [reply tests/unit/components/playback/controls/InputDropDown.spec.js:125]: Agree

File: tests/unit/components/playback/controls/InputDropDown.spec.js [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+    const parent_id_events = wrapper.emitted("update:value");
+    expect(parent_id_events).toBeUndefined();
+  });
+  test("When the InputDropDown is mounted, Then the widget width is modified in proption to that of the value set from the props value 'entry_width'", async () => {
```
[the remark this answers is not in the document]
Comment (robinr) [reply tests/unit/components/playback/controls/InputDropDown.spec.js:129]: Input-DropDown-200px
Input-DropDown-300

Why Proportion ?
So the width is known from a prop say either as 200px or 300px etc now the text in the title is also know from the title_label.. in terms of CSS property if suppose the text is just say "ID" is placed inside the <span> tag </span>
under span text centre automatically doesn't work it positions the text based on left: <value>px, which I have computed with a small math percentage compute so I have used proportion. 

In the correction now first I would be verifying if <div> or <p> or <header> tag computes centre and doesn't need a match for left: <value>px, hope fully one of them might have centre work on the text then, I will remove the compute and correct the word proportion. Will let you know the result.

[the remark this answers is not in the document]
Comment (robinr) [reply tests/unit/components/playback/controls/InputDropDown.spec.js:129]: some text got removed in between or its DIV or P or H  tags

File: tests/unit/components/playback/controls/InputDropDown.spec.js [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+      "width: 390px;"
+    );
+    const input_text_entry_feedback = wrapper.find(
+      ".div__input-dropdown-controls-content-feedback"
+    );
+    expect(input_text_entry_feedback.attributes("style")).toStrictEqual(
+      "width: 390px; top: 88px;"
+    );
+  });
+  test("When the InputDropDown is mounted, Then the widget width is modified in proption to that of the value set from the props value 'entry_width' in the event of title being empty the height is modified without a hole", async () => {
```
[the remark this answers is not in the document]
Comment (robinr) [reply tests/unit/components/playback/controls/InputDropDown.spec.js:174]: Agreed, would correct accordingly.

File: components/playback/controls/player/InputDropDown.vue [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+    message_if_blank: { type: Boolean, default: false }, // when set to true, will display a simple feedback
+  },
+  data() {
+    return {
+      input_dropdown_value_key: this.value,
+      input_width_background: this.input_width + 4,
+    };
+  },
+  computed: {
+    input_key_action: function () {
```
[the remark this answers is not in the document]
Comment (robinr) [reply components/playback/controls/player/InputDropDown.vue:95]: input_key_action function is not needed will remove.

File: components/playback/controls/player/InputDropDown.vue (hunk 1 above)
Comment (robinr) [reply components/playback/controls/player/InputDropDown.vue:100]: Agreed will try to introduce the unit test to validate the sequence.

**Comment by robinr** [comment]
Working on the above ... comments. would let you know.

File: tests/e2e/playback/controls/player/input-dropdown-e2e.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (eli88fine) [reply tests/e2e/playback/controls/player/input-dropdown-e2e.js:63]: why do you need to have a calculation based on the number of characters in the title label? can you use the same solution as the TextInput widget where the `span` holding the title is set to the same `left` and `width` as the dropdown menu itself and within the `span` the text is center-aligned?

File: tests/unit/components/playback/controls/InputDropDown.spec.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (eli88fine) [reply tests/unit/components/playback/controls/InputDropDown.spec.js:129]: In the title of your test, you have the word  `proption` , which is not a word in English. I was assuming you meant to say "proportion", but if you meant something else then write that instead. But `proption` is not a word

[the remark this answers is not in the document]
Comment (robinr) [reply tests/unit/components/playback/controls/InputDropDown.spec.js:129]: Correction is needed, would reword as  "props"

File: tests/e2e/playback/controls/player/input-dropdown-e2e.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (robinr) [reply tests/e2e/playback/controls/player/input-dropdown-e2e.js:63]: Yes true, corrected the CSS as left : 5px it wasn't centre, now set to left: 0px; as a result align : center; and 
text-align: center;  is placing the text in middle of the widget irrespective of size of the label string.

**Comment by robinr** [comment]
Complete all the PR review comments and changes.

File: tests/unit/components/playback/controls/InputDropDown.spec.js [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+      propsData,
+      store,
+      localVue,
+    });
+    const target_input = wrapper.find("#input-dropdown-widget");
+    expect(target_input.attributes().placeholder).toStrictEqual(
+      "Select Customer ID"
+    );
+  });
+  test("When the InputDropDown is mounted, Then the user enters few charters in the input, confirm that an event 'update' is emitted with entered text", async () => {
```
[the remark this answers is not in the document]
Comment (eli88fine) [reply tests/unit/components/playback/controls/InputDropDown.spec.js:67]: it does not look like this was addressed in your new commits

File: tests/unit/components/playback/controls/InputDropDown.spec.js [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+      propsData,
+      store,
+      localVue,
+    });
+    const target_div = wrapper.find(
+      ".div__input-dropdown-controls-content-feedback"
+    );
+    expect(target_div.text()).toStrictEqual("This field is required");
+  });
+  test("When the InputDropDown is mounted, Then the user enters few charters in the input, verify that its prevented as props have made this option disabled", async () => {
```
[the remark this answers is not in the document]
Comment (eli88fine) [reply tests/unit/components/playback/controls/InputDropDown.spec.js:106]: it does not look like this was addressed in your new commits

File: tests/unit/components/playback/controls/InputDropDown.spec.js [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+    const customer = "New Customer"; // new Customer
+    const input_widget = wrapper.find("#input-dropdown-widget");
+    input_widget.element.value = customer;
+    await input_widget.trigger("input");
+    const parent_id_events = wrapper.emitted("update:value");
+    expect(parent_id_events).toHaveLength(1);
+    expect(parent_id_events).toStrictEqual([[customer]]);
+  });
+  test("When the InputDropDown is mouted, Then a prop message_if_blank is set to true and invalid_text is provided as value is <empty>, an error text is rendered", async () => {
+    const propsData = {
```
Comment (eli88fine) [inline tests/unit/components/playback/controls/InputDropDown.spec.js:88]: please also assert that the `div__input-dropdown-controls-content-widget--invalid` class has been applied to the box, since that is also supposed to happen when it is invalid

File: tests/unit/components/playback/controls/InputDropDown.spec.js [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+    };
+    wrapper = mount(ComponentToTest, {
+      propsData,
+      store,
+      localVue,
+    });
+    const target_div = wrapper.find(
+      ".div__input-dropdown-controls-content-feedback"
+    );
+    expect(target_div.text()).toStrictEqual("This field is required");
```
Comment (eli88fine) [inline tests/unit/components/playback/controls/InputDropDown.spec.js:104]: It looks like if the field is optional (`message_if_blank=false`), then the error does not show up if a user types in an invalid choice.  Please create a failing unit test replicating this issue and then correct the problem in the code.
image

**Review by eli88fine — CHANGES_REQUESTED** [review]
good progress. a few items weren't addressed from the first set of comments. Also there's a scenario where the invalid coloring does not show up when it should.  

Also, it looks like you did not `rebase` from the `development` branch.  When you check to see whether or not the CI checks passed, there's a box next to it saying whether or not you need to `rebase`---right now it says "out of date with development branch" which means you need to rebase.
image

File: tests/unit/components/playback/controls/InputDropDown.spec.js (hunk 13 above)
Comment (robinr) [reply tests/unit/components/playback/controls/InputDropDown.spec.js:88]: Now asserted for visibility.

File: tests/unit/components/playback/controls/InputDropDown.spec.js (hunk 14 above)
Comment (robinr) [reply tests/unit/components/playback/controls/InputDropDown.spec.js:104]: NO don't agree, this is not problem, if some one wants to have the field message_if_black=false, he is forcing the system to not give feedback,  the logic of validation is under the parent or the user of this component. 

Incase , if there is no feedback to provide then message_if_black should be set to true and the invalid_text be blank.

When some one enter "**boetheant**" the parent has to set the value of message_if_black=false on then response receiving event function, has to run the logic of validation and set the message_if_black=true and give the feedback text in the invalid_text reason, this is not a part of InputDropDown Widget functionality.

JEST UNIT testcase can't have a logic of the parent validation brought into the widget verification or testing, this is not feasible.

Just like all the validation or managed in the parent for **InputWidget**, same happens in **InputDropDown**. 
So both have identical control, this is a wrong observation which cannot be implemented, in the InputDropDown widget.

Comment (eli88fine) [reply tests/unit/components/playback/controls/InputDropDown.spec.js:108]: Agreed---we can have this component be very flexible by not including any validation logic at all.

However, we have two very common use cases for the validation logic of a dropdown widget, so in the near future we should create a parent widget to make the use cases easy to code:

In almost all cases, we want something to be considered invalid if it does not match a choice from the list. We would use the same invalid message for any situation (except blank) where the user input does not match a choice in the list, the input would be valid if it matched any choice in the list.  So something with props like: `message_when_not_matching_list`, and `message_when_blank` (which could be an empty string indicating that the field is optional and blank is valid). And then that component takes care of that simple and very common validation logic

File: tests/unit/components/playback/controls/InputDropDown.spec.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (eli88fine) [reply tests/unit/components/playback/controls/InputDropDown.spec.js:67]: it still does not look like this was addressed in your new commits

File: tests/unit/components/playback/controls/InputDropDown.spec.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (eli88fine) [reply tests/unit/components/playback/controls/InputDropDown.spec.js:110]: it still does not look like this was addressed in your new commits

**Review by eli88fine — CHANGES_REQUESTED** [review]
I agree with your assessment of the validation logic---see comment for details.

There are two comments about test titles that have been noted since the first review that are still not addressed. Please look at those and then I can approve for merge

File: tests/unit/components/playback/controls/InputDropDown.spec.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (robinr) [reply tests/unit/components/playback/controls/InputDropDown.spec.js:67]: oh, there was small link called hidden, I didn't realise there were comments.  I will correct them.

Just now I clicked the text "hidden" and saw the comments.

File: tests/unit/components/playback/controls/InputDropDown.spec.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (robinr) [reply tests/unit/components/playback/controls/InputDropDown.spec.js:110]: oh, there was small link called hidden, I didn't realise there were comments.  I will correct them.

Just now I clicked the text "hidden" and saw the comments.

File: tests/unit/components/playback/controls/InputDropDown.spec.js [hunk 15]
```diff
@@ -64,7 +64,7 @@ describe("InputDropDown.vue", () => {
       "Select Customer ID"
     );
   });
-  test("When the InputDropDown is mounted, Then the user enters few charters in the input, confirm that an event 'update' is emitted with entered text", async () => {
+  test("Given that the user types, When a valid option from the list is entered, Then confirm that an event 'update' is emitted with entered text", async () => {
```
Comment (eli88fine) [inline tests/unit/components/playback/controls/InputDropDown.spec.js:67]: a "Given" section is not necessary in every test title. If there is nothing special about the way you are setting up the environment or mounting the component, then it is fine to just start with a "When" section.   

In this test, the user typing is the direct action that triggers the change we are trying to test. For example, something like: `When the user types a valid option from the list, Then an event 'update' is emitted with the entered text`

**Review by eli88fine — CHANGES_REQUESTED** [review]

**Review by eli88fine — CHANGES_REQUESTED** [review]
There still looks like some misinterpretation about what should be in the `When` vs the `Given` section of the test title. See comments---we can also discuss live on our call to clarify

**Review by eli88fine — APPROVED** [review]
